### ISSUE TITLE
Add a LocalRuntime for replacing Runtime+LocalSet

### ISSUE THREAD

**Issue by Darksonn** [tokio-rs/tokio#6739]
Currently, to spawn `!Send` tasks on Tokio you must use a combination of a current-thread runtime and `LocalSet` to make that work. However, tasks on a `LocalSet` are separate from the rest of the runtime in an uncomfortable way. For example:

* Tasks on the `LocalSet` can `tokio::spawn` to get onto the runtime, but once you're on the runtime, you can no longer `spawn_local` to get back into the `LocalSet`.
* From the runtime's perspective, all of the tasks in the `LocalSet` behave like a single task using `FuturesUnordered`, so various runtime options such as [`event_interval`](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.event_interval) behave in a very surprising way by counting many tasks at once.
* After discussions with Deno, the `LocalSet` has been shown to involve considerable performance overhead compared to unsafely spawning the `!Send` tasks onto a current-thread runtime.

For the above reasons, I propose introducing a new type called `LocalRuntime` which behaves exactly like a current-thread `Runtime` with the following exceptions:

* The `LocalRuntime` type is `!Send` to enable spawning `!Send` tasks directly onto the runtime.
* From within a `LocalRuntime`, the `tokio::spawn` and `spawn_local` functions have identical behavior.
* Tasks spawned using `spawn_local` are scheduled in the same way as normal `tokio::spawn` tasks.

Since `Runtime` and `LocalSet` share a fair amount of their internals, the implementation of current-thread should not require changes to support this.

**Comment by Noah-Kennedy** [comment]
I've opened a PR to add this as an unstable API: https://github.com/tokio-rs/tokio/pull/6808

**Comment by carllerche** [comment]
Generally, I agree with the reasoning for adding `LocalSet`.  We would need to think about how to position this with the current thread runtime. Both LocalRuntime and CurrentThreadRuntime have a lot of overlap. So, how to pick each one would need to be clear.

**Comment by Noah-Kennedy** [comment]
> Generally, I agree with the reasoning for adding `LocalSet`.  We would need to think about how to position this with the current thread runtime. Both LocalRuntime and CurrentThreadRuntime have a lot of overlap. So, how to pick each one would need to be clear.

I think the decision tree here comes down to several things:
- do you want to be able to seamlessly switch between multi_thread and current_thread (e.g via a config file entry in your service)
- do you want to move runtimes between threads (not as dumb as it sounds)
- do you actually care about spawn_local

**Comment by Sytten** [comment]
Any progress on stabilizing this? @Noah-Kennedy

**Comment by juchiast** [comment]
Can we stabilize this?

**Comment by lquerel** [comment]
Do we have an estimate of when `LocalRuntime` will be available without `RUSTFLAGS="--cfg tokio_unstable"`?

**Comment by Darksonn** [comment]
I think we can stabilize.

**Comment by ADD-SP** [comment]
Hello folks, tokio is moving forward with the stabilization of `LocalRuntime`, we welcome community involvement, especially in discussing downstream impacts.

See #7558

**Comment by strohel** [comment]
Shall this issue be closed in favor of https://github.com/tokio-rs/tokio/issues/7558? `LocalRuntime` exists in unstable tokio, and the stabilization is tracked in the linked issue.

**Comment by BatmanAoD** [comment]
> After discussions with Deno, the LocalSet has been shown to involve considerable performance overhead compared to unsafely spawning the !Send tasks onto a current-thread runtime.

Is there somewhere I can get more information about the performance implications of `LocalSet`? I'm seeing some very weird behavior that seems to be related to slow polling of a task running in a `LocalSet`, and although I'm going to go ahead and switch to a `LocalRuntime`, I would like to have a better idea of whether the issues I'm seeing are expected.

**Comment by Darksonn** [comment]
Ultimately it boils down to the fact that `LocalSet` is kind of similar to running everything in a single task using `FuturesUnordered`. For example, Tokio metrics will see the local tasks as one big combined task.

**Comment by DanglingPointer** [comment]
> Ultimately it boils down to the fact that `LocalSet` is kind of similar to running everything in a single task using `FuturesUnordered`. For example, Tokio metrics will see the local tasks as one big combined task.

Does that mean that if one of the futures/tasks is notified, every future will be polled?

**Comment by lquerel** [comment]
Do we have any idea when LocalRuntime will be in a stable state?

**Comment by ADD-SP** [comment]
> Do we have any idea when LocalRuntime will be in a stable state?

@lquerel I think there is a soft blocker for stabilization: <https://github.com/tokio-rs/tokio/pull/7420>. It should be ok to stabilize the `LocalRuntime` without it, but it would be better having it before stabilization.

**Comment by Darksonn** [comment]
> Does that mean that if one of the futures/tasks is notified, every future will be polled?

No, like `FuturesUnordered` it keeps track of which futures it makes sense to poll.

> Do we have any idea when LocalRuntime will be in a stable state?

I suggested stabilizing it some time ago, but some concerns were raised. I'm not sure if we worked through them.

**Comment by ADD-SP** [comment]
> I suggested stabilizing it some time ago, but some concerns were raised. I'm not sure if we worked through them.

Based on the tracking issue (<https://github.com/tokio-rs/tokio/issues/7558>), I think there is no **hard** blocker.

**Comment by lquerel** [comment]
@Darksonn @ADD-SP 

I'm not a contributor to this project, but as a heavy user of it - and of `LocalSet` in particular - I'm in favor of releasing `LocalRuntime` as it is :-)

**Comment by Darksonn** [comment]
Then let's do it.
